TEST: Add xfailing test for [func-returns-value]

harahu opened the pull request:
<!-- If this pull request fixes an issue, add "Fixes #NNN" with the issue number. -->

Add an xfailing test that reproduces #14179.

<!--
Checklist:
- Read the [Contributing Guidelines](https://github.com/python/mypy/blob/master/CONTRIBUTING.md)
- Add tests for all changed behaviour.
- If you can't add a test, please explain why and how you verified your changes work.
- Make sure CI passes.
- Please do not force push to the PR once it has been reviewed.
-->

harahu commented:
Is there a canonical way to point back to the github issue from the test code, so that if the test passes at some point, it is easy to also close the issue?

brianschubert commented:
> Is there a canonical way to point back to the github issue from the test code, so that if the test passes at some point, it is easy to also close the issue?

I don't know about a canonical way, but any comment that points to the issue number / URL would probably be fine.
There's some [existing test cases](https://github.com/search?q=repo%3Apython%2Fmypy++path%3Atest-data++%28%22see%22+OR+%22TODO%22+OR+%22issue%22%29&type=code) that do this in various ways.

harahu commented:
> > Is there a canonical way to point back to the github issue from the test code, so that if the test passes at some point, it is easy to also close the issue?
> 
> I don't know about a canonical way, but any comment that points to the issue number / URL would probably be fine. There's some [existing test cases](https://github.com/search?q=repo%3Apython%2Fmypy++path%3Atest-data++%28%22see%22+OR+%22TODO%22+OR+%22issue%22%29&type=code) that do this in various ways.

I added the URL as a comment.

brianschubert commented on test-data/unit/check-errorcodes.test:
```diff
@@ -563,6 +563,15 @@ y = A().g()  # E: "g" of "A" does not return a value (it only ever returns None)
+
+y1 = A.g()  # E: "g" of "A" does not return a value (it only ever returns None)  [func-returns-value]
+y2 = A().g()  # E: "g" of "A" does not return a value (it only ever returns None)  [func-returns-value]
+
```
```suggestion
[builtins fixtures/staticmethod.pyi]

```
Missing fixture for `staticmethod`.  See details at https://github.com/python/mypy/blob/master/test-data/unit/README.md#fixtures

brianschubert commented:
I think this test would be better suited for `check-expressions.test`. There's a block of similar cases with the prefix `testNoneReturn`.

`check-errorcodes.test` for tests related to error codes themselves.

brianschubert commented:
This issue appear to affect any decorated function, so there may be value in including a test using an arbitrary dummy decorator, and with multiple levels of decoration.

For your consideration, please feel free to adapt as you see fit:
```python
# test-data/unit/check-expressions.test
[case testNoneReturnDecorated]
from typing import Any, Callable, TypeVar

F = TypeVar('F', bound=Callable[..., Any])

def deco(f: F) -> F:
    pass

@deco
@deco
def f() -> None: 
    pass

class A:
    @staticmethod
    def s() -> None:
        pass

if int():
    x = f()      # E: "f" does not return a value (it only ever returns None)
if int():
    x = A.s()    # E: "s" of "A" does not return a value (it only ever returns None)
if int():
    x = A().s()  # E: "s" of "A" does not return a value (it only ever returns None)
[builtins fixtures/staticmethod.pyi]
```

harahu commented:
> I think this test would be better suited for `check-expressions.test`. There's a block of similar cases with the prefix `testNoneReturn`.
> 
> `check-errorcodes.test` is for tests related to error codes themselves.

Moved and renamed the test.

> This issue appear to affect any decorated function, so there may be value in including a test using an arbitrary dummy decorator, and with multiple levels of decoration.

Good catch! I expanded the test based on your example.

brianschubert reviewed (commented):
Nice

The `__call__` case is interesting. There are some other tests that cover opaque callables in other kinds of expressions, like `f = A(); y = f()`. It seems that the relevant bit here is having the callee be a `CallExpr`

A few nits on the error messages:

brianschubert reviewed (approved):
LGTM

Both tests pass when I force all function calls to always return `None`.

hauntsaninja commented:
Superseded by https://github.com/python/mypy/pull/18015
